Std/Data/UnionFind/Basic.lean

```diff
@@ -0,0 +1,438 @@
+/-
+Copyright (c) 2021 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Std.Data.Array.Lemmas
+
+namespace Std
+
+/-- Union-find node type -/
+structure UFNode where
+  /-- Parent of node -/
+  parent : Nat
+  /-- Rank of node -/
+  rank : Nat
+
+namespace UnionFind
+
+/-- Parent of a union-find node, defaults to self when the node is a root -/
+def parentD (arr : Array UFNode) (i : Nat) : Nat :=
+  if h : i < arr.size then (arr.get ⟨i, h⟩).parent else i
+
+/-- Rank of a union-find node, defaults to 0 when the node is a root -/
+def rankD (arr : Array UFNode) (i : Nat) : Nat :=
+  if h : i < arr.size then (arr.get ⟨i, h⟩).rank else 0
+
+theorem parentD_eq {arr : Array UFNode} {i} : parentD arr i.1 = (arr.get i).parent := dif_pos _
+
+theorem parentD_eq' {arr : Array UFNode} {i} (h) :
+    parentD arr i = (arr.get ⟨i, h⟩).parent := dif_pos _
+
+theorem rankD_eq {arr : Array UFNode} {i} : rankD arr i.1 = (arr.get i).rank := dif_pos _
+
+theorem rankD_eq' {arr : Array UFNode} {i} (h) : rankD arr i = (arr.get ⟨i, h⟩).rank := dif_pos _
+
+theorem parentD_of_not_lt : ¬i < arr.size → parentD arr i = i := (dif_neg ·)
+
+theorem lt_of_parentD : parentD arr i ≠ i → i < arr.size :=
+  Decidable.not_imp_comm.1 parentD_of_not_lt
+
+theorem parentD_set {arr : Array UFNode} {x v i} :
+    parentD (arr.set x v) i = if x.1 = i then v.parent else parentD arr i := by
+  rw [parentD]; simp [Array.get_eq_getElem, parentD]
+  split <;> [split <;> simp [Array.get_set, *]; split <;> [(subst i; cases ‹¬_› x.2); rfl]]
+
+theorem rankD_set {arr : Array UFNode} {x v i} :
+    rankD (arr.set x v) i = if x.1 = i then v.rank else rankD arr i := by
+  rw [rankD]; simp [Array.get_eq_getElem, rankD]
+  split <;> [split <;> simp [Array.get_set, *]; split <;> [(subst i; cases ‹¬_› x.2); rfl]]
+
+end UnionFind
+
+open UnionFind
+
+/-- Union-find data structure -/
+structure UnionFind where
+  /-- Array of union-find nodes -/
+  arr : Array UFNode
+  /-- Validity for parent nodes -/
+  parentD_lt : ∀ {i}, i < arr.size → parentD arr i < arr.size
+  /-- Validity for rank -/
+  rankD_lt : ∀ {i}, parentD arr i ≠ i → rankD arr i < rankD arr (parentD arr i)
```

Comment:
Could you add more documentation to the structure here?

I think it should include names for the operations to:
* Create an empty UnionFind.
* Add a new union find node.
* Assert two union find nodes are equal.
* Check if two union find nodes are equal.

I don't think we've consistently done this for other operations, but at least in this case, I found it a bit difficult to find those operations.

Reply:
The operations provided are:

1. `empty`: Create an empty UnionFind.
2. `push: Push a new UnionFind node not equal to anything else.
3. `find`: Find the canonical representative of a UnionFind node, and update the data structure to make subsequent lookups faster
4. `root`: Find the canonical representative of a UnionFind node without changing the structure
5. `union`: Join the equivalence classes of two UnionFind nodes.

Operations (3) and (5) give the data structure its name. The reason why this is not more abstracted is because *finding is a mutating operation*, and this fact is key to the performance of the data structure. You can use operation (4) to find instead of (3), but it would make the performance much worse and would be a performance pitfall because it has a nicer signature. Your two operations "Assert two union find nodes are equal" and "check if two union find nodes are equal" have possible implementations using (3) or (4), but only the version using (4) can be used for specification purposes and we don't want people to stumble into doing that instead of using (3).

Reply:
Let me know if I missed something.

Reply:
I moved `UnionFind.Equiv` to `Basic` even though its (deliberately) noncomputable. That just made more sense to me.